README.md
# DiscUnion

## Description

Discriminated unions for Elixir - for building algebraic data types.

Allows for building data structure with a closed set of representations/cases as
an alternative for a set of tuple+atom combo. Elixir already had product type -
tuples. With DiscUnion library, sum-types, types with a fixed set of values can
be created (also called discriminated unions or disjoint unions).

Provides macros and functions for creating and matching on datastructres which
throw compile-time and run-time exceptions if an unknow case was used or not all
cases were covered in a match. It's inspired by ML/OCaml/F# way of building
discriminated unions. Unfortunately, Elixir does not support such a strong
typing and this library will not solve this. However, it allows to easily catch
common mistakes at compile-time instead of run-time (those can be sometimes hard
to detect).

## How to use

(In `example` folder, there is a tennis kata example, a simple coding exercises,
that shows exactly how to use this library.)

To define a discriminated union, `defunion` macro is used. Use `|` to separate
union cases from each other. Union cases can have arguments and an asterisk `*`
can be used to combine several arguments. Underneath, it's just a struct with
union cases represented as atoms and tuples. Type specs in definitions are
passed to `@spec` declaration, so dialyzer can be used. However, DiscUnion does
not type-check anything by it self.

### Usage

``` elixir
defmodule Shape do
  use DiscUnion

  defunion Point
  | Circle in float()
  | Rectangle in any * any
end
```

Type specs in `Circle` or `Rectangle` definitions are only for description and
have no influence on code nor are they used for any type checking - there is no
typchecking other then checking if correct cases were used!

When constructing a case (an union tag), you have couple of options:

 * `c` macro, where arity depends on number of arguments you set for
   cases (compile-time checking),
 * `c!` function, where arity depends on number of arguments you set for
   cases (run-time checking),
 * `from/1` macro, accepts a tuple (compile-time checking),
 * `from!/` or `from!/2` functions, accepts a tuple (only run-time checking).
 * a dynamically built macro (aka "named constructors") named after union tag
   (in a camelized form, i.e. `Point`'s `Rectangle` case, would be available as
   `Point.rectangle/2` macro and also with compile-time checking),

Run-time constructors `from!` can be overridden. Any changes in functionality
introduced to them will also impact `c!` constructors which are based on
`from!`. This, for example, allows for defining variant cases with some run-time
validations.

Preferred way to construct a variant case is via `c` macros or `c!`
functions. `from/1` and `from!/1` construcotrs are mainly to be used when
interacting with return values like in example with opening a file. If you'd
like to enable named constructors do:
`use DiscUnion, named_constructors: true`.

If `Score.from {Pointz, 1, 2}` or `Score.c Pointz, 1, 2`, from tennis kata
example, be placed somewhere in `run_test_match/0` function compiler would throw
this error:

``` elixir
== Compilation error on file example/tennis_kata.exs ==
** (UndefinedUnionCaseError) undefined union case: Pointz in _, _
    (disc_union) expanding macro: Score.from/1
    (disc_union) example/tennis_kata.exs:38: Tennis.run_test_match/0
```

If you would use `from!/1` or `c!`, this error would be thrown at run-time, or,
in the case of `from!/2`, not at all! Function `from!/2` returns it's second
argument when unknow clause is passed to the function.


For each discriminated union, a special `case` macro is created. This macro
checks if all cases were covered in it's clauses (at compile-time) and expects
it's predicate to be evaluated to this discriminated union's struct (checked at
run-time).

If `Game in _`, in `Tennis.score_point/2` functions, would be commented,
compiler would throw this error:

``` elixir
== Compilation error on file example/tennis_kata.exs ==
** (MissingUnionCaseError) not all defined union cases are used, should be all of: Points in "PlayerPoints" * "PlayerPoints", Advantage in "Player", Deuce, Game in "Player"
    (disc_union) expanding macro: Score.case/2
    (disc_union) example/tennis_kata.exs:64: Tennis.score_point/2
```

You can also use a catch-all statement (_), like in a regular `case` macro
(`Kernel.SpecialForms.case/2`), but here, it needs to be explicitly enabled by
passing `allow_underscore: true` option to the macro:

``` elixir
Score.case score, allow_underscore: true do
  Points in PlayerPoints.forty, PlayerPoints.forty -> Score.duce
  _ -> score
end
```

Otherwise you would see a smillar error like above.


## How it works

Underneath, it's just a module containg a struct with tuples and some
dynamically built macros. This property can be used for matching in function
definitions, although it will not look as clearly as a `case` macro built for a
discriminated union.


The `Shape` union creates a `%Shape{}` struct with current active case held in
`case` field and all possible cases can be get by `Shape.__union_cases__/0`
function:

``` elixir
%Shape{case: Point} = Shape.c Point
%Shape{case: {Circle, :foo}} = Shape.c Circle, :foo
```

Cases that have arguments are just tuples; *n*-argument union case is a
*n+1*-tuple with a case tag as it's first element. This should work seamlessly
with existing conventions:

``` elixir
defmodule Result do
  use DiscUnion

  defunion :ok in any | :error in atom
end

defmodule Test do
  use Result

  def run(file) do
    res = Result.from! File.open(file)
    Result.case res do
      r={:ok, io_dev}                       -> {:yey, r, io_dev}
      :error in reason when reason==:eacces -> :too_much_protections
      :error in :enoent                     -> :why_no_file
      :error in _reason                     -> :ney
    end
  end
end
```
Since cases are just a tuples, they can be also used as a clause for `case`
macro. Matching and gaurds also works!

### Side note

It is possible to place discriminated union's constructor macros in function
definition:

``` elixir
defmodule ShapeArea do
  use   Shape

  def calc_area(Shape.c(Point)), do: 0
  def calc_area(Shape.circle(r)), do: :math.pi*r*r  # assuming named construcors are enabled
end
```

And even use natural Elixir's multi-fun capability, to build logic, like in
`score_point/2` function in tennis kata example. Although, placing constructors
inside of function definition is not a bad thing, and being able to do so is a
clear WIN! Using this technique to build business logic, instead of using
discriminated union's `case` macro, is not encouraged because nothing checks if
all union cases were covered.


## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
as:

  1. Add disc_union to your list of dependencies in `mix.exs`:

        def deps do
          [{:disc_union, "~> 0.3.0"}]
        end

  2. Ensure disc_union is started before your application:

        def application do
          [applications: [:disc_union]]
        end


CHANGELOG.md
# Changelog

## v0.3.0

  * Update to Elixir 1.7+
  * Exception raised by run-time constructors contain correct case name & count
  * Make run-time constructors overridable

## v0.2.0

  * Improve docs
  * Add `c` & `c!` constructors
  * Add named constructors
  * Generate docs & typespecs for discriminated unions
  * Raise errors with clean stacktrace

## v0.1.0

  * Initial working version


test/example_du.ex
defmodule ExampleDU do
  use DiscUnion, named_constructors: true
  defunion Asd
  | Qwe in any
  | Rty in integer * atom
end


test/example_dua.ex
defmodule ExampleDUa do
  use DiscUnion, named_constructors: true
  defunion :asd
  | :qwe in any
  | :rty in integer * atom
end


mix.exs
defmodule DiscUnion.Mixfile do
  use Mix.Project

  def project do
    [
      app: :disc_union,
      version: "0.3.0",
      elixir: "~> 1.2",
      description: "Discriminated unions for Elixir - for building algebraic data types",
      package: package(),
      compilers: compilers(),
      test_coverage: [tool: ExCoveralls],
      preferred_cli_env: [coveralls: :test, "coveralls.html": :test,
                          "coveralls.detail": :test, "coveralls.post": :test],
      consolidate_protocols: Mix.env != :test,
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(),
      elixirc_paths: elixirc_paths(Mix.env),
      docs: docs()
    ]
  end

  # Configuration for the OTP application
  #
  # Type "mix help compile.app" for more information
  def application do
    [applications: [:logger]]
  end

  defp deps do
    [
      {:excoveralls, "~> 0.4", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 0.5", only: [:dev], runtime: false},
      {:ex_doc, "~> 0.21", only: [:dev], runtime: false},
      {:earmark, "~> 1.4", only: [:dev], runtime: false},
      {:credo, "~> 1.1", only: [:dev], runtime: false}
    ]
  end

  defp compilers do
    compilers(Mix.env)
  end
  defp compilers(_), do: Mix.compilers

  defp package do
    [ maintainers: ["X4lldux"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/x4lldux/disc_union"},  ]
  end

  defp elixirc_paths(:test), do: ["lib", "test"]
  defp elixirc_paths(_), do: ["lib"]

  defp docs do
    [extras: [
        "README.md": [title: "README"],
        "CHANGELOG.md": [title: "Changelog"]
      ]]
  end
end


lib/utils.ex
defmodule DiscUnion.Utils do
  @moduledoc false

  @type canonical_union_tag :: {:_, 0} | {:__aliases__, atom} | atom

  @spec canonical_form_of_union_case(Macro.expr | {atom, any} | any) :: {canonical_union_tag, non_neg_integer, any}
  def canonical_form_of_union_case(c) do
    case c do
      {:{}, _ctx, [union_tag | union_args]} ->
        str_case = [union_tag | union_args]
        |> Enum.map(&Macro.to_string/1)
        |> List.to_tuple
        {union_tag |> canonical_union_tag, union_args |> length, str_case}
      {union_tag, union_arg} ->
        str_case = [union_tag, union_arg]
        |> Enum.map(&Macro.to_string/1)
        |> List.to_tuple
        {union_tag |> canonical_union_tag, 1, str_case}
      union_tag ->
        {union_tag |> canonical_union_tag, 0, c |> Macro.to_string}
    end
  end

  @spec canonical_union_tag(Macro.expr) :: canonical_union_tag
  defp canonical_union_tag({:_, _, _}), do: {:_, 0}
  defp canonical_union_tag({:__aliases__, _, union_tag}), do: {:__aliases__, union_tag}
  defp canonical_union_tag(union_tag), do: union_tag

  @spec extract_union_case_definitions(Macro.expr) :: [Macro.t]
  def extract_union_case_definitions(x) do
    extract_union_case_definitions(x, []) |> Enum.reverse
  end
  def extract_union_case_definitions({:in, _, [left, right]}, acc) do
    vars = right |> unstar

    [{:{}, [], [left | vars]} | acc]
  end
  def extract_union_case_definitions({:|, _, [left, right]}, acc) do
    extract_union_case_definitions(right, extract_union_case_definitions(left, acc) )
  end
  def extract_union_case_definitions(other, acc) do
    [other | acc]
  end

  @spec unstar(Macro.expr) :: [Macro.t]
  defp unstar(expr) do
    unstar(expr, []) |> Enum.reverse
  end
  defp unstar({:*, _, [left, right]}, acc) do
    unstar(right, unstar(left, acc))
  end
  defp unstar(other, acc) do
    [other |> Macro.to_string | acc]
  end

  @spec is_cases_valid?([Macro.expr]) :: :ok | {:error, :not_atoms} | {:error, :not_unique}
  def is_cases_valid?(cases) do
    cond do
      not is_only_atoms? cases -> {:error, :not_atoms}
      not is_unique? cases -> {:error, :not_unique}
      true -> :ok
    end
  end

  @spec is_only_atoms?([Macro.expr]) :: boolean
  def is_only_atoms?(cases) do
    cases
    |> Enum.all?(
                 fn
                   {:__aliases__, _, [x|_]} when is_atom x               -> true
                   {:{}, _, [{:__aliases__, _, [x|_]}|_]} when is_atom x -> true
                   {:{}, _, [x|_]} when is_atom x                        -> true
                   {{:__aliases__, _, [x|_]}, _} when is_atom x          -> true
                   {x, _} when is_atom x                                 -> true
                   x when is_atom x                                      -> true
                   _                                                     -> false
                 end
    )
  end

  @spec is_unique?([Macro.expr]) :: boolean
  def is_unique?(cases) do
    unique_cases = cases
    |> Enum.map(fn
      {:__aliases__, _, [x|_]} when is_atom x                                  -> x
      {:{}, _, [{:__aliases__, _, [x|_]}|_]} when is_atom x                    -> x
      {:{}, _, [x|_]} when is_atom x                                           -> x
      {{:__aliases__, _, [x|_]}, _} when is_tuple(x) and elem(x, 0) |> is_atom -> x |> elem(0)
      {x, _} when is_atom x                                                    -> x # not really needed,
      # after transormation of cases, 2-tuples are converted in to n-tuples with two elemtns (in AST terms)
      # it's here just for symmetry with `is_only_atoms?/1`, which needs it, cause it's used in one other
      # place also.
      x when is_atom x                                                          -> x
    end)
    |> Enum.uniq

    length(cases) == length(unique_cases)
  end

  @spec module_name(atom) :: String.t
  def module_name(module) do
    case module |> to_string do
      "Elixir." <> m -> m
      m              -> m
    end
  end

  @spec build_union_case_spec(Macro.t) :: Macro.t
  def build_union_case_spec(variant_case) do
    canonical_case =
      variant_case
      |> DiscUnion.Utils.canonical_form_of_union_case

    case canonical_case do
      {_tag, _count, str} when  is_tuple(str) ->
        tuple_elems =
          str
          |> Tuple.to_list
          |> Enum.map(& String.replace(&1, "\"", "") )
          |> Enum.map(fn e -> Code.eval_string("quote do #{e} end", [], __ENV__) |> elem(0) end)
        quote do {unquote_splicing(tuple_elems)} end
      {_tag, _count, str} ->
        str
        |> Code.eval_string([], __ENV__)
        |> elem(0)
    end
  end

  @spec build_union_cases_specs(Macro.t) :: Macro.t
  def build_union_cases_specs(cases) do
    specs = cases
    |> Enum.map(&build_union_case_spec/1)
    |> Enum.reverse
    |> Enum.reduce(fn e, acc ->
        quote do unquote(e) | unquote(acc) end
    end)
    quote do
      @type t() :: %__MODULE__{case: unquote(specs)}
    end
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :disc_union, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:disc_union, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/exceptions.ex
defmodule UndefinedUnionCaseError do
  defexception [case: nil, case_args_count: 0]

  def message(exception = %{case_args_count: 0}) do
    "undefined union case: #{exception.case}"
  end
  def message(exception) do
    case_args = 0..exception.case_args_count-1
    |> Enum.map(fn _ -> "_" end)
    |> Enum.join(" * ")
    "undefined union case: #{exception.case} in #{case_args}"
  end
end

defmodule MissingUnionCaseError do
  defexception [cases: nil]

  defp format_case(c) when is_tuple(c) do
    [tag | args] = c
    |> Tuple.to_list

    args = args |> Enum.join(" * ")
    "#{tag} in #{args}"
  end
  defp format_case(c), do: c |> to_string

  def message(exception) do
    if "_" in exception.cases do
      cases = exception.cases
      |> Enum.map(&format_case/1)
      |> Enum.reject(&match?("_", &1))
      |> Enum.join(", ")

      "not all defined union cases are used, should be at least a catch all statement (_) and any combination of: #{cases}"
    else
      cases = exception.cases
      |> Enum.map(&format_case/1)
      |> Enum.join(", ")
      "not all defined union cases are used, should be all of: #{cases}"
    end
  end
end


lib/utils/case.ex
defmodule DiscUnion.Utils.Case do
  @moduledoc false

  alias DiscUnion.Utils

  @type case_clause :: {:->, [{atom, any}], [any]}
  @type case_clauses :: [case_clause]

  defmacro raise_undefined_union_case(c, at: :runtime) do
    quote location: :keep, bind_quoted: [c: c] do
      try do
        raise "oops"
      rescue
        exception ->
          stacktrace = System.stacktrace() |> Enum.drop(1)

          {_, union_args_count, str_form} = c |> Utils.canonical_form_of_union_case
          union_tag = case str_form do
                        x when is_tuple(x) -> x |> elem(0)
                        x  -> x
                      end
          reraise UndefinedUnionCaseError, [case: union_tag, case_args_count: union_args_count], stacktrace
        end
    end
  end

  defmacro raise_undefined_union_case(c, at: :compiletime) do
    quote location: :keep, bind_quoted: [c: c] do
      try do
        raise "oops"
      rescue
        exception ->
          stacktrace = Enum.drop_while(System.stacktrace(),
            fn {_, _, _, o} ->
              file = Keyword.get(o, :file) |> to_string
              String.contains?(file, __ENV__.file |> Path.basename)
              or
              String.contains?(file, "enum.ex") # HACK: ugggh! Somebody please suggest a better way
            end)
            |> Enum.drop(1)

          {_, union_args_count, str_form} = c |> Utils.canonical_form_of_union_case
          union_tag = case str_form do
                        x when is_tuple(x) -> x |> elem(0)
                        x  -> x
                      end
          reraise UndefinedUnionCaseError, [case: union_tag, case_args_count: union_args_count], stacktrace
        end
    end
  end


  @spec raise_missing_union_case(case_clauses) :: no_return()
  def raise_missing_union_case(all_cases) do
    try do
      raise "oops"
    rescue
      exception ->
        stacktrace = System.stacktrace
        if Exception.message(exception) == "oops" do
          stacktrace = Enum.drop_while(stacktrace,
            fn {_, _, _, o} ->
              Keyword.get(o, :file)
              |> to_string
              |> String.contains?(__ENV__.file |> Path.basename)
            end)
          cases = all_cases |> Enum.map(&elem(&1, 2))
          reraise MissingUnionCaseError, [cases: cases], stacktrace
        end
    end
  end

  @spec transform_case_clauses(case_clauses | nil, term, boolean) :: case_clauses | no_return()
  # when `:do` is empty
  def transform_case_clauses(nil, all_union_cases, _allow_underscore) do
    raise_missing_union_case all_union_cases
  end
  def transform_case_clauses(clauses, all_union_cases, allow_underscore) do
    underscore_canonical_case = Macro.var(:_, nil) |> Utils.canonical_form_of_union_case
    all_union_cases = case allow_underscore do
                        true  -> all_union_cases ++ [underscore_canonical_case]
                        false -> all_union_cases
                      end

    {clauses, simplified_used_cases} = normalize_and_simplify_used(clauses,
      all_union_cases)

    underscore_base_case_form = base_case_form underscore_canonical_case
    if (length all_union_cases) > (length simplified_used_cases) &&
      not(underscore_base_case_form in simplified_used_cases) do
      raise_missing_union_case all_union_cases
    end

    clauses
  end

  defp normalize_and_simplify_used(clauses, all_union_cases) do
    simplifier = fn {:->, ctx, [clause | clause_body]}, simplified_used_cases ->
      # normalizes format when clauses are in `in` format or are 2-tuple
      {transformed_clause, _} = map_reduce_clauses(clause, &transform_case_clause/2, [])

      # raise on unknown case clauses
      map_reduce_clauses(transformed_clause,
        &check_for_unknown_case_clauses/2,
        {ctx, all_union_cases})

      # simplified used cases for check later
      {_, simplified_used_cases} = map_reduce_clauses(transformed_clause,
        &extract_used_case_clauses/2,
        simplified_used_cases)

    {{:->, ctx, [transformed_clause | clause_body]}, simplified_used_cases}
    end

    Enum.map_reduce(clauses, [], simplifier)
  end

  # transforms case macro clauses to a common format, if they are in `in` format or they are 2-tuples
  defp transform_case_clause([{:in, ctx, [union_tag | [union_arg] ]} | rest_of_union_args], acc) do
    elems = [union_tag, union_arg | rest_of_union_args]
    {[{:{}, ctx, elems}], acc}
  end
  defp transform_case_clause([{union_tag = {_, ctx, _}, union_args}], acc) do
    elems = [union_tag, union_args]
    {[{:{}, ctx, elems}], acc}
  end
  defp transform_case_clause(c, acc) do
    {c, acc}
  end

  @spec check_for_unknown_case_clauses(Macro.t, any) :: {Macro.t, any}
  defp check_for_unknown_case_clauses([c], acc = {_ctx, all_cases}) do
    known? =
      c
      |> Utils.canonical_form_of_union_case
      |> is_case_clause_known?(all_cases)
    if known? do
      :ok
    else
      raise_undefined_union_case(c, at: :compiletime)
    end

    {[c], acc}
  end

  defp is_case_clause_known?(canonical_form, all_cases) do
    {canonical_union_tag, canonical_union_args_count, _} = canonical_form
    Enum.any?(all_cases, fn {tag, args_count, _} ->
      {canonical_union_tag, canonical_union_args_count} == {tag, args_count}
    end)
  end

  defp extract_used_case_clauses([c], used_cases) do
    base_case_form = base_case_form c
    used_cases = case base_case_form in used_cases do
      false -> [base_case_form | used_cases]
      true  -> used_cases
    end

    {[c], used_cases}
  end

  # case form without string representation
  defp base_case_form(c) do
    {canonical_union_tag, canonical_union_args_count, _} = Utils.canonical_form_of_union_case c
    {canonical_union_tag, canonical_union_args_count}
  end

  @spec map_reduce_clauses(case_clauses, (list(Macro.expr), any -> {Macro.expr, any}), any) :: {case_clauses, any}
  def map_reduce_clauses(clauses, f, acc) do
    {clauses, {_, acc}} = clauses |> map_reduce_clause({f, acc})
    {clauses, acc}
  end

  defp map_reduce_clause([{:=, ctx, [bind, precond]} | rest_of_union_args], {f, acc}) do
    {[precond|_], f_acc} =
      [precond | rest_of_union_args]
      |> map_reduce_clause({f, acc})

    {[{:=, ctx, [bind, precond]}], f_acc}
  end

  defp map_reduce_clause([{:when, ctx, [precond | guards_and_union_args]}], {f, acc}) do
    guard = guards_and_union_args |> List.last
    union_arg = guards_and_union_args |> List.delete_at(-1)

    {[precond|_], f_acc} =
      [precond | union_arg]
      |> map_reduce_clause({f, acc})

    {[{:when, ctx, [precond, guard]}], f_acc}
  end

  defp map_reduce_clause(elem, {f, acc}) do
    {new_elem, f_acc} = f.(elem, acc)
    {new_elem, {f, f_acc}}
  end

end


lib/disc_union.ex
defmodule DiscUnion do
  @moduledoc """
  Discriminated unions for Elixir - for building algebraic data types.

  Allows for building data structure with a closed set of representations/cases as
  an alternative for a set of tuple+atom combo. Elixir already had product type -
  tuples. With DiscUnion library, sum-types, types with a fixed set of values can
  be created (also called discriminated unions or disjoint unions).

  Provides macros and functions for creating and matching on datastructres which
  throw compile-time and run-time exceptions if an unknow case was used or not all
  cases were covered in a match. It's inspired by ML/OCaml/F# way of building
  discriminated unions. Unfortunately, Elixir does not support such a strong
  typing and this library will not solve this. However, it allows to easily catch
  common mistakes at compile-time instead of run-time (those can be sometimes hard
  to detect).

  To use it, you need to add:  `use DiscUnion` to your module.

  ## How it works

  Underneath, it's just a module containg a struct with tuples and some
  dynamically built macros. This property can be used for matching in function
  definitions, although it will not look as clearly as a `case` macro built for a
  discriminated union.


  The `Shape` union creates a `%Shape{}` struct with current active case held in
  `case` field and all possible cases can be get by `Shape.__union_cases__/0`
  function:

  ``` elixir
  %Shape{case: Point} = Shape.c Point
  %Shape{case: {Circle, :foo}} = Shape.c Circle, :foo
  ```

  Cases that have arguments are just tuples; *n*-argument union case is a
  *n+1*-tuple with a case tag as it's first element. This should work seamlessly
  with existing conventions:

  ``` elixir
  defmodule Result do
    use DiscUnion

    defunion :ok in any | :error in atom
  end

  defmodule Test do
    use Result

    def run(file) do
      res = Result.from! File.open(file)
      Result.case res do
        r={:ok, io_dev}                       -> {:yey, r, io_dev}
        :error in reason when reason==:eacces -> :too_much_protections
        :error in :enoent                     -> :why_no_file
        :error in _reason                     -> :ney
      end
    end
  end
  ```
  Since cases are just a tuples, they can be also used as a clause for `case`
  macro. Matching and gaurds also works!
  """

  defmacro __using__(opts) do
    opts = opts ++ [named_constructors: false]
    if true == Keyword.get opts, :named_constructors do
      Module.put_attribute(__CALLER__.module, :named_constructors, true)
    end

    quote do
      require DiscUnion
      require DiscUnion.Utils.Case
      require DiscUnion.Utils.Constructors

      import DiscUnion, only: [defunion: 1]
    end
  end

  @doc """
  Defines a discriminated union.

  To define a discriminated union, `defunion` macro is used. Use `|` to separate
  union cases from each other. Union cases can have arguments and an asterisk
  `*` can be used to combine several arguments. Underneath, it's just a struct
  with union cases represented as atoms and tuples. Type specs in definitions
  are passed to `@spec` declaration, so dialyzer can be used. However, DiscUnion
  does not type-check anything by it self.

  ### Usage

  ``` elixir
  defmodule Shape do
    use DiscUnion

    defunion Point
    | Circle in float()
    | Rectangle in any * any
  end
  ```

  Type specs in `Circle` or `Rectangle` definitions are only for description and
  have no influence on code nor are they used for any type checking - there is
  no typchecking other then checking if correct cases were used!

  When constructing a case (an union tag), you have couple of options:

   * `c` macro, where arity depends on number of arguments you set for
     cases (compile-time checking),
   * `c!` function, where arity depends on number of arguments you set for
     cases (run-time checking),
   * `from/1` macro, accepts a tuple (compile-time checking),
   * `from!/` or `from!/2` functions, accepts a tuple (only run-time checking).
   * a dynamically built macro (aka "named constructors") named after union tag
     (in a camelized form, i.e. `Point`'s `Rectangle` case, would be available
     as `Point.rectangle/2` macro and also with compile-time checking),

  Run-time constructors `from!` can be overridden. Any changes in functionality
  introduced to them will also impact `c!` constructors which are based on
  `from!`. This, for example, allows for defining variant cases with some
  run-time validations.

  Preferred way to construct a variant case is via `c` macros or `c!`
  functions. `from/1` and `from!/1` construcotrs are mainly to be used when
  interacting with return values like in example with opening a file. If you'd
  like to enable named constructors do:
  `use DiscUnion, named_constructors: true`.

  If `Score.from {Pointz, 1, 2}` or `Score.c Pointz, 1, 2`, from tennis kata
  example, be placed somewhere in `run_test_match/0` function compiler would
  throw this error:

  ``` elixir
  == Compilation error on file example/tennis_kata.exs ==
  ** (UndefinedUnionCaseError) undefined union case: Pointz in _, _
      (disc_union) expanding macro: Score.from/1
      (disc_union) example/tennis_kata.exs:38: Tennis.run_test_match/0
  ```

  If you would use `from!/1` or `c!`, this error would be thrown at run-time,
  or, in the case of `from!/2`, not at all! Function `from!/2` returns it's
  second argument when unknow clause is passed to the function.


  For each discriminated union, a special `case` macro is created. This macro
  checks if all cases were covered in it's clauses (at compile-time) and expects
  it's predicate to be evaluated to this discriminated union's struct (checked
  at run-time).

  If `Game in _`, in `Tennis.score_point/2` functions, would be commented,
  compiler would throw this error:

  ``` elixir
  == Compilation error on file example/tennis_kata.exs ==
  ** (MissingUnionCaseError) not all defined union cases are used, should be all of: Points in "PlayerPoints" * "PlayerPoints", Advantage in "Player", Deuce, Game in "Player"
      (disc_union) expanding macro: Score.case/2
      (disc_union) example/tennis_kata.exs:64: Tennis.score_point/2
  ```

  You can also use a catch-all statement (_), like in a regular `case` macro
  (`Kernel.SpecialForms.case/2`), but here, it needs to be explicitly enabled by
  passing `allow_underscore: true` option to the macro:

  ``` elixir
  Score.case score, allow_underscore: true do
    Points in PlayerPoints.forty, PlayerPoints.forty -> Score.duce
    _ -> score
  end
  ```

  Otherwise you would see a smillar error like above.
  """
  defmacro defunion(expr) do
    cases = DiscUnion.Utils.extract_union_case_definitions(expr)

    Module.register_attribute __CALLER__.module, :cases_canonical, persist: true
    Module.put_attribute(__CALLER__.module,
                         :cases_canonical,
                         cases |> Enum.map(&DiscUnion.Utils.canonical_form_of_union_case/1))

    case DiscUnion.Utils.is_cases_valid? cases do
      {:error, :not_atoms} -> raise ArgumentError, "union case tag must be an atom"
      {:error, :not_unique} -> raise ArgumentError, "union case tag must be unique"
      :ok -> build_union cases
    end
  end

  defp build_union(cases) do
    union_typespec = DiscUnion.Utils.build_union_cases_specs(cases)
    main_body = quote location: :keep, unquote: true do
      all_cases = unquote(cases)
      @enforce_keys [:case]

      defstruct case: []

      defimpl Inspect do
        import Inspect.Algebra

        def inspect(union, opts) do
          mod = @for |> Module.split
          concat ["##{mod}<", Inspect.inspect(union.case, opts), ">"]
        end
      end

      @doc "Returns a list with all acceptable union cases."
      def __union_cases__ do
        unquote(cases)
      end

      def __using__(_) do
        quote do
          require unquote(__MODULE__)
        end
      end

      @doc """
      Matches the given expression against the given clauses. The expressions
      needs to be evaluate to `%#{DiscUnion.Utils.module_name __MODULE__}{}`.
      """
      defmacro case(expr, do: block) do
                 do_case expr, [], do: block
               end
      defmacro case(expr, [allow_underscore: true], do: block) do
                 do_case expr, [allow_underscore: true], do: block
               end
      defmacro case(expr, opts, do: block) do
                 do_case expr, [], do: block
               end

      @spec do_case(Macro.t, Keyword.t, [do: Macro.t]) :: Macro.t
      defp do_case(expr, opts, do: block) do
        opts = opts ++ [allow_underscore: false]
        mod = __MODULE__
        allow_underscore = Keyword.get opts, :allow_underscore
        block = DiscUnion.Utils.Case.transform_case_clauses(block,
          @cases_canonical, allow_underscore)
        quote location: :keep do
          precond = unquote expr
          mod = unquote mod
          if not match?(%{__struct__: mod}, precond) do
            raise BadStructError, struct: mod, term: precond
          end
          case precond.case do
            unquote(block)
          end
        end
      end

      DiscUnion.Utils.Constructors.build_constructor_functions __MODULE__, all_cases
    end

    [union_typespec, main_body]
  end
end
